### PR TITLE
Add tests for ipython notebooks to run during Travis CI

### PR DESCRIPTION
**Pull request by mon95**
Added tests for ipython notebooks to run during CI and check each cell of the notebooks for correct execution. Errors due to timeout resulting from wget commands are handled. If the dataset is not available then the ipynb is skipped. 
Solves https://github.com/RaRe-Technologies/gensim/issues/936

### CODE REVIEW

**Comment by tmylk** [comment]
The `wget` or other timeouts should not be silent. Add raising of an explicit 'SKIP' to the ipynb if it happens.

**Comment by tmylk** [comment]
Plus there are CI failures in Travis.

File: gensim/test/test_notebooks.py [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+def _notebook_run(path):
+    """
+    Execute a notebook via nbconvert and collect output.
+       :returns (parsed nb object, execution errors)
+    """
+    kernel_name = 'python%d' % sys.version_info[0]
+    this_file_directory = os.path.dirname(__file__)
+    errors = []
+
+    fl = tempfile.NamedTemporaryFile(suffix=".ipynb", mode='wt', delete=False)
```
Comment (mon95) [inline gensim/test/test_notebooks.py]: There is no need for it while testing. Will remove it

File: gensim/test/test_notebooks.py [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+
+import logging
+import unittest
+import os
+import os.path
+import tempfile
+import numbers
+import codecs   
+
+class TimeoutError(Exception):
```
Comment (mon95) [inline gensim/test/test_notebooks.py]: Removing it gives me this error on timeout: 
```NameError: global name 'TimeoutError' is not defined```

Is this a python2 vs python3 issue? I understand that TimeoutError is a subclass of OSError but I'm not sure how to use it in code.

**Comment by mon95** [comment]
Will modify the ipython notebooks to include something like this:
```raise ValueError(\"SKIP: please download the questions-word.txt```

Would this be alright? This would get detected as a CellExecutionError.

File: gensim/test/test_notebooks.py (hunk 2 above)
Comment (tmylk) [inline gensim/test/test_notebooks.py]: It is a Python 3 error. Why does it need special treatment?

**Comment by tmylk** [comment]
Adding SKIP instead of download is ok. It would be best to make sure it only raises it during testing.

**Comment by mon95** [comment]
Sure, I'll make the necessary changes to the ipython notebooks.

File: gensim/test/test_notebooks.py (hunk 2 above)
Comment (tmylk) [inline gensim/test/test_notebooks.py]: why should we catch any of them at all?

Comment (mon95) [inline gensim/test/test_notebooks.py]: Changed this. Thank you!

**Comment by gojomo** [comment]
Won't this change the CI tests from taking about 5 minutes to run, to many hours? 

Also, notebooks sometimes include code that's intended to error (as illustration) or examples of variants (not intended to be always run)... so this automatically-run-everything approach seems constraining to me.

File: docs/notebooks/Similarity_Queries.ipynb [hunk 3]
```diff
@@ -60,8 +60,11 @@
    "source": [
     "from gensim import corpora, models, similarities\n",
     "\n",
-    "dictionary = corpora.Dictionary.load('/tmp/deerwester.dict')\n",
-    "corpus = corpora.MmCorpus('/tmp/deerwester.mm') # comes from the first tutorial, \"From strings to vectors\"\n",
+    "try:\n",
+    "    dictionary = corpora.Dictionary.load('/tmp/deerwester.dict')\n",
+    "    corpus = corpora.MmCorpus('/tmp/deerwester.mm') # comes from the first tutorial, \"From strings to vectors\"\n",
+    "except:\n",
+    "    raise ValueError(\"SKIP: Dependency - Corpora_and_Vector_Spaces.ipynb file\")\n",
```
Comment (tmylk) [inline docs/notebooks/Similarity_Queries.ipynb]: actually these 2 notebooks should be united into one as they are dependent.

File: docs/notebooks/online_w2v_tutorial.ipynb [hunk 4]
```diff
@@ -78,7 +79,10 @@
    },
    "outputs": [],
    "source": [
-    "old, new = [WikiCorpus('enwiki-{}-pages-articles.xml.bz2'.format(ymd)) for ymd in ['20101011', '20160820']]"
+    "try:\n",
+    "    old, new = [WikiCorpus('enwiki-{}-pages-articles.xml.bz2'.format(ymd)) for ymd in ['20101011', '20160820']]\n",
+    "except:\n",
+    "    raise ValueError(\"SKIP: please download the wiki dump files.\")\n"
```
Comment (tmylk) [inline docs/notebooks/online_w2v_tutorial.ipynb]: Please download the wiki dump files by uncommenting the wget commands above.

File: docs/notebooks/Similarity_Queries.ipynb (hunk 3 above)
Comment (mon95) [inline docs/notebooks/Similarity_Queries.ipynb]: Moved the Similarity queries tutorial to the Corpora_and_Vector_Spaces.ipynb file. 
Will change links within tutorials.md if this is alright.

**Comment by mon95** [comment]
@gojomo The tests shouldn't take too long to execute as we're setting each cell's execution time limit to be about 10 seconds(this can be changed).

**Comment by gojomo** [comment]
@mon95 So when a cell times out – or the next cell depends on the results of timed-out cell – the notebook registers as a test failure? (While I haven't looked at all the notebooks, the ones I'm familiar with tend to have multiple cells that take more than 10 seconds to execute.)

**Comment by mon95** [comment]
As of now, some of the ipython notebooks have been modified to include something like this:
```raise ValueError(\"SKIP: please download the questions-word.txt``` 

And these CellExecutionErrors do not lead to a test failure.

**Comment by gojomo** [comment]
@mon95 That seems different than timeout failures, and I don't see similar changes in notebooks like doc2vec-IMDB.ipynb, where many cells need tens-of-minutes to complete. Is the intent that this notebooks-as-tests only runs on some subset of `docs/notebooks`, those that are intended to run straight-through quickly?

**Comment by mon95** [comment]
Haven't run into errors or timeouts on those notebooks yet because other errors seem to crop up while testing. Will make necessary changes to the notebooks if such timeouts are encountered. 

Notebooks that require large datasets that are skipped. The idea is to get the tests to run in CI. https://github.com/RaRe-Technologies/gensim/issues/936

**Comment by tmylk** [comment]
@mon95 do the tests run on your local machine?

**Comment by mon95** [comment]
Yes! The tests run for both python2.7 and python3.5

**Comment by tmylk** [comment]
Please add nbconvert to [test dependencies](https://github.com/RaRe-Technologies/gensim/blob/develop/setup.py#L183). It will probably solve the import problems

**Comment by mon95** [comment]
Added nbconvert to the test dependencies.

**Comment by tmylk** [comment]
@mon95 Tests fail because nbconvert can't be installed in travis for some reason. Can you see Travis logs?

**Comment by mon95** [comment]
Yes! I'm able to see the logs now. I'll try the solution that's mentioned in the last comment here [https://github.com/jupyter/jupyter/issues/150](url)

**Comment by tmylk** [comment]
@olavurmortensen Your new LDA tutorial doesn't pass the CI on notebooks - please add data if it is small, or add a SKIP option.

**Comment by tmylk** [comment]
The goal of this PR is to make CI tests pass in at least one version of Python and skip on others. It will already be a great improvement.
Next stage will be to make them pass in other Python versions.

**Comment by tmylk** [comment]
Need to add pyldavis to [test dependencies](https://github.com/RaRe-Technologies/gensim/blob/9a3329cfccfcf98e9b88f5f52437f1049f4e3c44/setup.py#L183)

**Comment by mon95** [comment]
@tmylk Could you please take a look at this? 
The tests for the notebooks do seem to run fine and they fail elsewhere (TestLdaMulticore).

**Comment by markroxor** [comment]
@mon95 please resolve the conflicts.
@tmylk Are you available to review this? or should I?

**Comment by tmylk** [comment]
Looking for volunteer to resolve the conflicts, then should be good to merge after a quick review.

**Comment by tmylk** [comment]
Suggest using nbdime package https://github.com/jupyter/nbdime

**Pull request closed without merging**